src/etl/normalizer.py

```diff
@@ -0,0 +1,223 @@
+        pattern = re.compile(rf"\(\s*({valid_units})\s*\)")
+        return self.convert_units(df_base, valid_units, pattern, "(inches)")
+
+    def normalize_pound_columns(self, df_base: pl.DataFrame) -> pl.DataFrame:
+        """
+        Converts column values that may come in units other than pound.
+        It converts from grams, kilograms and once to pounds.
+        """
+        valid_units = "|".join(self.data_config["pounds_units"])
+        pattern = re.compile(rf"\(\s*({valid_units})\s*\)")
+        return self.convert_units(df_base, valid_units, pattern, "(pounds)")
+
+    def convert_units_to_float(self, df_base: pl.DataFrame) -> pl.DataFrame:
+        """
+        Converts the inches and pound columns to float.
+        """
+        columns_to_parse = df_base.select(
+            r"^.*\(\s*(inches|pounds)\s*\).*$"
+        ).columns
+        dimension_pattern = re.compile(r".*dimensions?.*")
+        for col in columns_to_parse:
+            if dimension_pattern.search(col):
+                continue
+            df_base = df_base.with_columns(pl.col(col).cast(pl.Float64))
+
+        return df_base
+
+    def convert_currency(self, df_base: pl.DataFrame) -> pl.DataFrame:
+        """
+        Get all the currency columns, replace the unwanted characters
+        and format with 2 decimal places.
+        """
+
+        # Regex pattern to get all the columns that has the $ symbol.
+        column_name_pattern = r"^[^$]*\(\$\)[^$]*$"
+        currency_columns = df_base.select(column_name_pattern).columns
+        values_pattern = r"[^\d.]"
+
+        return df_base.with_columns(
+            [
+                pl.col(col)
+                .fill_null("0.00")
+                .str.replace_all(values_pattern, "")
+                .cast(pl.Float64)
+                .map_elements(lambda x: f"{x:.2f}", return_dtype=pl.String)
+                for col in currency_columns
+            ]
+        )
+
+    def filter_wrong_upcs(self, df_base: pl.DataFrame) -> pl.DataFrame:
+        """
+        It filters out the upcs that are null and saves the resulting dataframe.
+        We can then report these missing upcs to the corresponding client.
+        """
+        mask_wrong_upc = pl.col("upc").is_null()
+
+        wrong_upc_quantity = df_base.filter(mask_wrong_upc).shape[1]
+        self.logger.warning(
+            f"[NORMALIZER] Removed {wrong_upc_quantity} null upcs."
+        )
+
+        return df_base.filter(~mask_wrong_upc)
+
+    def convert_upc_to_ean13(self, df_base: pl.DataFrame) -> pl.DataFrame:
+        """
+        It converts the upc column to ean13 column.
+        """
+        self.logger.info("[NORMALIZER] Converting upc to ean13.")
+        return df_base.with_columns(
+            pl.format(
+                "0{}-{}-{}",
+                pl.col("upc").str.slice(0, length=2),
+                pl.col("upc").str.slice(2, length=9),
+                pl.col("upc").str.slice(-1),
+            ).alias("upc")
+        )
```

Comment:
Nice overall, but you could 'gold plate' this solution by validating the check digit.

https://en.wikipedia.org/wiki/Check_digit#UPC,_EAN,_GLN,_GTIN,_numbers_administered_by_GS1